dnsimple/dnsimple.go

```diff
@@ -51,7 +44,19 @@ type Client struct {
 
 // NewClient returns a new DNSimple API client.
 func NewClient(apiToken, email string) *Client {
-	c := &Client{ApiToken: apiToken, Email: email, HttpClient: &http.Client{}, BaseURL: baseURL, UserAgent: userAgent}
+	credentials := NewApiTokenCredentials(email, apiToken)
+	c := &Client{Credentials: credentials, HttpClient: &http.Client{}, BaseURL: baseURL, UserAgent: userAgent}
+	c.Contacts = &ContactsService{client: c}
+	c.Domains = &DomainsService{client: c}
+	c.Registrar = &RegistrarService{client: c}
+	c.Users = &UsersService{client: c}
+	return c
+}
+
+// NewAuthenticatedClient returns a new DNSimple API client  using the given
+// credentials.
+func NewAuthenticatedClient(credentials Credentials) *Client {
```

Comment:
I'm not thrilled by the idea of duplicating the initialization of the API client. It will force us to add two lines, one for each method, each time we add a new service handler.

I see two options here. The first one is that `NewClient` sets the credential and then calls `NewAuthenticatedClient` passing the default credential instance.

The other option, which is the one I probably prefer, is that we change `NewClient` to accept a `credential` value rather than `apiToken` and `email`, and we remove the `NewAuthenticatedClient`.

The reason I prefer the second one is because it's cleaner (less code) and because there may be cases where it's legit to call an endpoint without being authenticated. It also seems more flexible to me because we extract the signature (and requirements) of the credential elsewhere, rather than coupling the client with the specific credential type.